File: node/src/providers/credentials.ts
import type { EndpointProfileId, EndpointProtocol, ProviderId } from "./endpoints.js"

export type ProviderCredential =
  | { type: "api_key"; value: string }
  | { type: "bearer"; value: string }

export interface CredentialRequest {
  providerId: ProviderId
  modelId: string
  endpointId: EndpointProfileId
  protocol: EndpointProtocol
}

export type CredentialResolver = (
  request: CredentialRequest,
) => ProviderCredential | undefined | Promise<ProviderCredential | undefined>

export interface OAuthAccessToken {
  accessToken: string
  /** Epoch time on the clock supplied to the resolver. The token is invalid at this instant. */
  expiresAt: number
  scopes?: readonly string[]
  audience?: string
}

export interface OAuthRefreshRequest {
  providerId: ProviderId
  audience?: string
  requiredScopes: readonly string[]
}

export interface OAuthCredentialResolverOptions {
  providerId: ProviderId
  /** Provider-owned refresh/ADC bridge. The Host owns its transport and refresh token. */
  refresh: (request: OAuthRefreshRequest) => OAuthAccessToken | Promise<OAuthAccessToken>
  requiredScopes?: readonly string[]
  audience?: string
  clock?: () => number
}

export interface CredentialOptions {
  apiKey?: string
  bearerToken?: string
  credentialResolver?: CredentialResolver
}

/** A deliberately non-sensitive credential failure. Never add resolver output to this error. */
export class CredentialResolutionError extends Error {
  readonly code:
    | "credential_unavailable"
    | "credential_invalid"
    | "credential_resolver_failed"
    | "credential_refresh_failed"
    | "credential_oauth_scope_mismatch"
    | "credential_oauth_audience_mismatch"
    | "credential_revoked"
    | "credential_auth_mode_unsupported"
  readonly providerId: ProviderId
  readonly retryable: boolean

  constructor(
    code: CredentialResolutionError["code"],
    providerId: ProviderId,
    retryable = false,
  ) {
    super(messageForCredentialError(code, providerId))
    this.name = "CredentialResolutionError"
    this.code = code
    this.providerId = providerId
    this.retryable = retryable
  }
}

/**
 * Host-owned OAuth/ADC extension. It retains only a refreshable bearer in process memory,
 * never exposes refresh-token material, and coalesces concurrent refreshes for one provider.
 */
export class OAuthCredentialResolver {
  private readonly providerId: ProviderId
  private readonly refresh: OAuthCredentialResolverOptions["refresh"]
  private readonly requiredScopes: readonly string[]
  private readonly audience?: string
  private readonly clock: () => number
  private token?: OAuthAccessToken
  private refreshInFlight?: Promise<OAuthAccessToken>
  private revoked = false

  constructor(options: OAuthCredentialResolverOptions) {
    this.providerId = options.providerId
    this.refresh = options.refresh
    this.requiredScopes = [...(options.requiredScopes ?? [])]
    this.audience = options.audience
    this.clock = options.clock ?? Date.now
  }

  /** An arrow field intentionally remains bound when passed as `credentialResolver`. */
  readonly resolve: CredentialResolver = async request => {
    if (request.providerId !== this.providerId) throw unavailable(request.providerId)
    if (this.revoked) throw new CredentialResolutionError("credential_revoked", this.providerId)

    const token = this.usable(this.token) ? this.token : await this.refreshToken()
    this.assertUsable(token)
    if (this.revoked) throw new CredentialResolutionError("credential_revoked", this.providerId)
    return { type: "bearer", value: token.accessToken }
  }

  revoke(): void {
    this.revoked = true
    this.token = undefined
  }

  status(): { providerId: ProviderId; revoked: boolean; hasUsableToken: boolean } {
    return {
      providerId: this.providerId,
      revoked: this.revoked,
      hasUsableToken: !this.revoked && this.usable(this.token),
    }
  }

  private usable(token: OAuthAccessToken | undefined): token is OAuthAccessToken {
    return token !== undefined && Number.isFinite(token.expiresAt) && token.expiresAt > this.clock()
  }

  private async refreshToken(): Promise<OAuthAccessToken> {
    if (!this.refreshInFlight) {
      this.refreshInFlight = Promise.resolve()
        .then(() => this.refresh({
          providerId: this.providerId,
          ...(this.audience === undefined ? {} : { audience: this.audience }),
          requiredScopes: this.requiredScopes,
        }))
        .then(token => {
          this.assertUsable(token)
          if (this.revoked) throw new CredentialResolutionError("credential_revoked", this.providerId)
          this.token = token
          return token
        })
        .catch(error => {
          if (error instanceof CredentialResolutionError) throw error
          throw new CredentialResolutionError("credential_refresh_failed", this.providerId, true)
        })
        .finally(() => { this.refreshInFlight = undefined })
    }
    return this.refreshInFlight
  }

  private assertUsable(token: OAuthAccessToken): void {
    if (
      !token
      || typeof token.accessToken !== "string"
      || token.accessToken.trim() === ""
      || !Number.isFinite(token.expiresAt)
      || token.expiresAt <= this.clock()
    ) {
      throw new CredentialResolutionError("credential_invalid", this.providerId)
    }
    if (this.audience !== undefined && token.audience !== this.audience) {
      throw new CredentialResolutionError("credential_oauth_audience_mismatch", this.providerId)
    }
    const tokenScopes = new Set(token.scopes ?? [])
    if (this.requiredScopes.some(scope => !tokenScopes.has(scope))) {
      throw new CredentialResolutionError("credential_oauth_scope_mismatch", this.providerId)
    }
  }
}

/** Safe for diagnostics and event metadata. It intentionally has no credential value. */
export function redactCredential(credential: ProviderCredential): { type: ProviderCredential["type"] } {
  return { type: credential.type }
}

export async function resolveCredential(
  request: CredentialRequest,
  options: CredentialOptions,
): Promise<ProviderCredential> {
  const configured = configuredCredential(request.providerId, options)
  if (configured) return configured
  if (!options.credentialResolver) throw unavailable(request.providerId)

  let resolved: ProviderCredential | undefined
  try {
    resolved = await options.credentialResolver(request)
  } catch (error) {
    if (error instanceof CredentialResolutionError) throw error
    throw new CredentialResolutionError("credential_resolver_failed", request.providerId)
  }
  return validateCredential(resolved, request.providerId)
}

/**
 * The legacy synchronous factory remains synchronous. A resolver which performs I/O must use
 * `resolveProviderRuntimeAsync` / `createProviderAsync` instead of risking an implicit request.
 */
export function resolveCredentialSync(
  request: CredentialRequest,
  options: CredentialOptions,
): ProviderCredential {
  const configured = configuredCredential(request.providerId, options)
  if (configured) return configured
  if (!options.credentialResolver) throw unavailable(request.providerId)

  let resolved: ProviderCredential | undefined | Promise<ProviderCredential | undefined>
  try {
    resolved = options.credentialResolver(request)
  } catch (error) {
    if (error instanceof CredentialResolutionError) throw error
    throw new CredentialResolutionError("credential_resolver_failed", request.providerId)
  }
  if (isPromise(resolved)) {
    throw new CredentialResolutionError("credential_resolver_failed", request.providerId)
  }
  return validateCredential(resolved, request.providerId)
}

function configuredCredential(
  providerId: ProviderId,
  options: CredentialOptions,
): ProviderCredential | undefined {
  const sources = [
    options.apiKey === undefined ? undefined : { type: "api_key" as const, value: options.apiKey },
    options.bearerToken === undefined ? undefined : { type: "bearer" as const, value: options.bearerToken },
  ].filter((value): value is ProviderCredential => value !== undefined)
  if (sources.length > 1) throw new CredentialResolutionError("credential_invalid", providerId)
  return sources.length === 1 ? validateCredential(sources[0], providerId) : undefined
}

function validateCredential(
  credential: ProviderCredential | undefined,
  providerId: ProviderId,
): ProviderCredential {
  if (!credential) throw unavailable(providerId)
  if (
    (credential.type !== "api_key" && credential.type !== "bearer")
    || typeof credential.value !== "string"
    || credential.value.trim() === ""
  ) {
    throw new CredentialResolutionError("credential_invalid", providerId)
  }
  return credential
}

function unavailable(providerId: ProviderId): CredentialResolutionError {
  return new CredentialResolutionError("credential_unavailable", providerId)
}

function messageForCredentialError(
  code: CredentialResolutionError["code"],
  providerId: ProviderId,
): string {
  switch (code) {
    case "credential_unavailable":
      return `Missing credential for provider ${providerId}`
    case "credential_invalid":
      return `Invalid credential for provider ${providerId}`
    case "credential_refresh_failed":
      return `Credential refresh failed for provider ${providerId}`
    case "credential_oauth_scope_mismatch":
      return `Credential scope does not satisfy provider ${providerId}`
    case "credential_oauth_audience_mismatch":
      return `Credential audience does not match provider ${providerId}`
    case "credential_revoked":
      return `Credential was revoked for provider ${providerId}`
    case "credential_auth_mode_unsupported":
      return `Credential authentication mode is unsupported for provider ${providerId}`
    case "credential_resolver_failed":
      return `Credential resolution failed for provider ${providerId}`
  }
}

function isPromise(value: unknown): value is Promise<unknown> {
  return Boolean(value) && typeof (value as { then?: unknown }).then === "function"
}

File: python/deepstrike/providers/credentials.py
"""Host-owned provider credential resolution, including the OAuth extension chain.

No credential object in this module is Kernel data.  OAuth refresh transport and any refresh
token remain in the caller-supplied callback; this module retains only the current access token.
"""
from __future__ import annotations

import asyncio
import inspect
import time
from dataclasses import dataclass
from typing import Awaitable, Callable, Literal, TypedDict


CredentialKind = Literal["api_key", "bearer"]


class CredentialResolutionError(Exception):
    """Secret-free, structured credential failure for the provider Host boundary."""

    def __init__(self, code: str, provider_id: str, *, retryable: bool = False) -> None:
        self.code = code
        self.provider_id = provider_id
        self.retryable = retryable
        super().__init__(_message(code, provider_id))


class OAuthAccessToken(TypedDict, total=False):
    access_token: str
    expires_at: float
    scopes: list[str]
    audience: str


OAuthRefresh = Callable[[], OAuthAccessToken | Awaitable[OAuthAccessToken]]


@dataclass(frozen=True)
class ProviderCredential:
    type: CredentialKind
    value: str


class CredentialRequest(TypedDict):
    provider_id: str
    model_id: str
    endpoint_id: str
    protocol: str


CredentialResolver = Callable[
    [CredentialRequest],
    ProviderCredential | None | Awaitable[ProviderCredential | None],
]


class OAuthCredentialResolver:
    """Refreshes a provider-owned bearer token once per expiry window.

    ``refresh`` is intentionally a narrow, Host-owned callback.  It receives no Kernel state and
    returns an access-token snapshot only; refresh tokens, SDK clients and HTTP mechanics never
    cross this boundary.
    """

    def __init__(
        self,
        *,
        provider_id: str,
        refresh: OAuthRefresh,
        required_scopes: tuple[str, ...] = (),
        audience: str | None = None,
        clock: Callable[[], float] | None = None,
    ) -> None:
        self._provider_id = provider_id
        self._refresh = refresh
        self._required_scopes = tuple(required_scopes)
        self._audience = audience
        self._clock = clock or time.time
        self._token: OAuthAccessToken | None = None
        self._refresh_in_flight: asyncio.Task[OAuthAccessToken] | None = None
        self._revoked = False

    async def resolve(
        self,
        *,
        provider_id: str,
        model_id: str,
        endpoint_id: str,
        protocol: str,
    ) -> str:
        del model_id, endpoint_id, protocol
        if provider_id != self._provider_id:
            raise CredentialResolutionError("credential_unavailable", provider_id)
        if self._revoked:
            raise CredentialResolutionError("credential_revoked", self._provider_id)
        token = self._token if self._is_usable(self._token) else await self._refresh_token()
        self._validate(token)
        if self._revoked:
            raise CredentialResolutionError("credential_revoked", self._provider_id)
        return token["access_token"]

    def revoke(self) -> None:
        self._revoked = True
        self._token = None

    def status(self) -> dict[str, object]:
        return {
            "provider_id": self._provider_id,
            "revoked": self._revoked,
            "has_usable_token": not self._revoked and self._is_usable(self._token),
        }

    def _is_usable(self, token: OAuthAccessToken | None) -> bool:
        return bool(
            token
            and isinstance(token.get("expires_at"), (int, float))
            and not isinstance(token.get("expires_at"), bool)
            and token["expires_at"] > self._clock()
        )

    async def _refresh_token(self) -> OAuthAccessToken:
        if self._refresh_in_flight is None:
            self._refresh_in_flight = asyncio.create_task(self._run_refresh())
        try:
            return await self._refresh_in_flight
        finally:
            if self._refresh_in_flight is not None and self._refresh_in_flight.done():
                self._refresh_in_flight = None

    async def _run_refresh(self) -> OAuthAccessToken:
        try:
            token = self._refresh()
            if inspect.isawaitable(token):
                token = await token
            self._validate(token)
            if self._revoked:
                raise CredentialResolutionError("credential_revoked", self._provider_id)
            self._token = token
            return token
        except CredentialResolutionError:
            raise
        except Exception as exc:
            raise CredentialResolutionError(
                "credential_refresh_failed", self._provider_id, retryable=True
            ) from None

    def _validate(self, token: OAuthAccessToken | None) -> None:
        if not token or not isinstance(token.get("access_token"), str) or not token["access_token"].strip():
            raise CredentialResolutionError("credential_invalid", self._provider_id)
        if not self._is_usable(token):
            raise CredentialResolutionError("credential_invalid", self._provider_id)
        if self._audience is not None and token.get("audience") != self._audience:
            raise CredentialResolutionError("credential_oauth_audience_mismatch", self._provider_id)
        scopes = token.get("scopes") or []
        if not isinstance(scopes, list) or any(scope not in scopes for scope in self._required_scopes):
            raise CredentialResolutionError("credential_oauth_scope_mismatch", self._provider_id)


async def resolve_credential(
    *,
    provider_id: str,
    model_id: str,
    endpoint_id: str,
    protocol: str,
    api_key: str | None = None,
    bearer_token: str | None = None,
    credential_resolver: CredentialResolver | OAuthCredentialResolver | None = None,
) -> ProviderCredential:
    configured = [
        ProviderCredential("api_key", api_key) if api_key is not None else None,
        ProviderCredential("bearer", bearer_token) if bearer_token is not None else None,
    ]
    configured = [credential for credential in configured if credential is not None]
    if len(configured) > 1:
        raise CredentialResolutionError("credential_invalid", provider_id)
    if configured:
        credential = configured[0]
        if not credential.value.strip():
            raise CredentialResolutionError("credential_invalid", provider_id)
        return credential
    if credential_resolver is None:
        raise CredentialResolutionError("credential_unavailable", provider_id)
    request: CredentialRequest = {
        "provider_id": provider_id,
        "model_id": model_id,
        "endpoint_id": endpoint_id,
        "protocol": protocol,
    }
    try:
        if isinstance(credential_resolver, OAuthCredentialResolver):
            resolved: ProviderCredential | None = ProviderCredential(
                "bearer",
                await credential_resolver.resolve(**request),
            )
        else:
            resolved = credential_resolver(request)
            if inspect.isawaitable(resolved):
                resolved = await resolved
    except CredentialResolutionError:
        raise
    except Exception:
        raise CredentialResolutionError("credential_resolver_failed", provider_id) from None
    return _validate_credential(resolved, provider_id)


def redact_credential(credential: ProviderCredential) -> dict[str, CredentialKind]:
    """Return diagnostic-safe credential metadata without retaining its value."""
    _validate_credential(credential, "unknown")
    return {"type": credential.type}


def _validate_credential(
    credential: ProviderCredential | None,
    provider_id: str,
) -> ProviderCredential:
    if credential is None:
        raise CredentialResolutionError("credential_unavailable", provider_id)
    if (
        not isinstance(credential, ProviderCredential)
        or credential.type not in {"api_key", "bearer"}
        or not isinstance(credential.value, str)
        or not credential.value.strip()
    ):
        raise CredentialResolutionError("credential_invalid", provider_id)
    return credential


def _message(code: str, provider_id: str) -> str:
    messages = {
        "credential_unavailable": "Missing credential",
        "credential_invalid": "Invalid credential",
        "credential_refresh_failed": "Credential refresh failed",
        "credential_resolver_failed": "Credential resolver failed",
        "credential_oauth_scope_mismatch": "Credential scope does not satisfy",
        "credential_oauth_audience_mismatch": "Credential audience does not match",
        "credential_revoked": "Credential was revoked",
        "credential_auth_mode_unsupported": "Credential authentication mode is unsupported for",
    }
    return f"{messages.get(code, 'Credential resolution failed')} provider {provider_id}"
